zio/parquetio/iterator.go

```diff
@@ -0,0 +1,726 @@
+)
+
+// Helper to avoid sprinkling fmt.Printf() calls around the code.
+// Enable debug statements by uncommenting the call to Printf()
+func debugf(msg string, args ...interface{}) {
+	//fmt.Printf(msg, args...)
+}
+
+// valueReader is the interface used for anything that can iterate over
+// a series of values of one of the parquet primitive types.  Note that
+// not all methods can be used on all implementations of this interface.
+type valueReader interface {
+	nextBoolean() bool
+	nextInt32() int32
+	nextInt64() int64
+	nextFloat() float64
+	nextDouble() float64
+	nextByteArray() []byte
+}
+
+// columnIterator reads and emits all the values from an individual
+// column inside a parquet file.  This implementation is not yet complete,
+// it only handles PLAIN and PLAIN_DICTIONARY encodings for a few
+// primitive data types.
+type columnIterator struct {
+	pr   *reader.ParquetReader
+	name string
+
+	maxRL int32
+	maxDL int32
+
+	rowGroup int
+
+	groupRead    int
+	groupTotal   int
+	thriftReader *thrift.TBufferedTransport
+	colMetadata  *parquet.ColumnMetaData
+
+	pageRead  int
+	pageTotal int
+
+	rlReader  *hybridReader
+	dlReader  *hybridReader
+	valReader valueReader
+
+	// only used for PageType_DICTIONARY_PAGE
+	int64Dict     []int64
+	floatDict     []float64
+	byteArrayDict [][]byte
+	// XXX need other types
+}
+
+func newColumnIterator(pr *reader.ParquetReader, name string, maxRL, maxDL int32) *columnIterator {
+	return &columnIterator{
+		pr:    pr,
+		name:  name,
+		maxRL: maxRL,
+		maxDL: maxDL,
+	}
+}
+
+func (i *columnIterator) clearDictionaries() {
+	i.byteArrayDict = nil
+	i.int64Dict = nil
+	i.floatDict = nil
+	// XXX clear others as they are added
+}
+
+// loadOnePage reads the next page that is part of this column.
+func (i *columnIterator) loadOnePage() (*parquet.PageHeader, []byte, error) {
+	// If we've reached the end of a row group (or if we haven't
+	// yet read the first row group), find the right offset inside
+	// the parquet file for this column in the next row group.
+	if i.groupRead == i.groupTotal {
+		if i.rowGroup >= len(i.pr.Footer.RowGroups) {
+			return nil, nil, io.EOF
+		}
+		rg := i.pr.Footer.RowGroups[i.rowGroup]
+		i.rowGroup++
+
+		i.groupRead = 0
+		i.groupTotal = int(rg.NumRows)
+
+		var col *parquet.ColumnChunk
+		for _, c := range rg.Columns {
+			if c.MetaData.PathInSchema[0] == i.name {
+				col = c
+				break
+			}
+		}
+		if col == nil {
+			return nil, nil, fmt.Errorf("cannot find ColumnChunk for %s", i.name)
+		}
+
+		if col.FilePath != nil {
+			return nil, nil, fmt.Errorf("ColumnChunk for %s refers to external file", i.name)
+		}
+
+		i.colMetadata = col.MetaData
+		offset := i.colMetadata.DataPageOffset
+		if i.colMetadata.DictionaryPageOffset != nil {
+			offset = *i.colMetadata.DictionaryPageOffset
+		}
+		size := i.colMetadata.TotalCompressedSize
+
+		// XXX
+		i.thriftReader = source.ConvertToThriftReader(i.pr.PFile, offset, size)
+
+		i.clearDictionaries()
```

Comment:
Shouldn't this get called whenever `loadDictionaryPage` is called?

Reply:
I don't think it matters, loadDictionaryPage overwrites any existing dictionary.  And any given column can't have dictionaries of different types.  This would be a lot nicer/clearer if we had generics rather than putting dictionaries of all the primitive types into this class...